longlong.h

```diff
@@ -399,17 +456,21 @@
   } while (0)
 
 #endif
+#endif
 
+#if !defined(__arm64__)
 #define sub_dddmmmsss(dh, dm, dl, mh, mm, ml, sh, sm, sl)           \
   do {                                                              \
     mp_limb_t __t, __u;                                             \
     sub_ddmmss(__t, dl, (mp_limb_t) 0, ml, (mp_limb_t) 0, sl);      \
     sub_ddmmss(__u, dm, (mp_limb_t) 0, mm, (mp_limb_t) 0, sm);      \
     sub_ddmmss(dh, dm, (mh) - (sh), dm, -__u, -__t);                \
   } while (0)
+#endif
 
 /* MIPS and ARM - Use clz builtins */
-#if (defined (__mips__) || defined (__arm__))
+/* NOTE: Apple clang version 12.0.5 miscompiles the count_leading_zeros fallback */
+#if (defined (__mips__) || defined (__arm__) || defined (__arm64__))
```

Comment:
Instead of hardcoding architectures and hoping that all compilers for these support certain builtins (they won't in general!) I would recommend to follow the example set by autoconf a couple decades ago and check in your configure script whether the compiler supports certain features or not (this does not require `autoconf`, of course).

Reply:
We gave up supporting other compilers ages ago. This file is not used on Windows at all, and we assume GCC or Clang which has not been a problem in practice. Very few other compilers support inline assembly in gnu as syntax anyway.

Reply:
As I see it, the problem is not the compiler, it is the hard coded list of architectures being checked here.

As such, I don't think this is so much about gnu as, as it is about using a compiler builtin as a fallback but only in a hard coded set of cases: it seems arbitrary to only use `__builtin_clzll` resp. `__builtin_clzll` to implement `count_leading_zeros` only on these architectures. Perhaps there is a deeper reason I am missing?

Indeed, if you are truly content to "require" gcc or clang (or anything "compatible") you could probably just *always* use those builtins (for GCC they have been there since *at least* 4.0; haven't checked for clang but I am pretty sure it implemented them long ago, too). In fact you could perhaps even remove the inline assembly implementations: they should compile to the same code (but of course that should be carefully verified first)

Reply:
Also, of course builtins can have bugs, so just because they were around back then doesn't necessarily mean they worked correctly, too, though I'd be hopeful. But of course that is a valid reason to be a bit wary.